plugins/guests/linux/cap/mount_virtualbox_shared_folder.rb

```diff
@@ -6,62 +6,85 @@ module VagrantPlugins
+            output = {stdout: '', stderr: ''}
+            uid_command = "id -u #{options[:owner]}"
+            machine.communicate.execute(uid_command,
+              error_class: Vagrant::Errors::VirtualBoxMountFailed,
+              error_key: :virtualbox_mount_failed,
+              command: uid_command,
+              output: output[:stderr]
+            ) { |type, data| output[type] << data if output[type] }
+            mount_uid = output[:stdout].chomp
+            @@logger.debug("Owner user ID (lookup): #{options[:owner]} -> #{mount_uid}")
           end
 
-          if options[:group].is_a? Integer
+          if options[:group].to_i.to_s == options[:group].to_s
             mount_gid = options[:group]
-            mount_gid_old = options[:group]
+            @@logger.debug("Owner group ID (provided): #{mount_gid}")
           else
-            mount_gid = "`getent group #{options[:group]} | cut -d: -f3`"
-            mount_gid_old = "`id -g #{options[:group]}`"
+            begin
+              output = {stdout: '', stderr: ''}
+              gid_command = "getent group #{options[:group]}"
+              machine.communicate.execute(gid_command,
+                error_class: Vagrant::Errors::VirtualBoxMountFailed,
+                error_key: :virtualbox_mount_failed,
+                command: gid_command,
+                output: output[:stderr]
+              ) { |type, data| output[type] << data if output[type] }
+              mount_gid = output[:stdout].split(':').at(2)
+              @@logger.debug("Owner group ID (lookup): #{options[:group]} -> #{mount_gid}")
+            rescue Vagrant::Errors::VirtualBoxMountFailed
+              if options[:owner] == options[:group]
+                @@logger.debug("Failed to locate group `#{options[:group]}`. Group name matches owner. Fetching effective group ID.")
+                output = {stdout: ''}
+                result = machine.communicate.execute("id -g #{options[:owner]}",
+                  error_check: false
+                ) { |type, data| output[type] << data if output[type] }
+                mount_gid = output[:stdout].chomp if result == 0
+                @@logger.debug("Owner group ID (effective): #{mount_gid}")
+              end
+              raise unless mount_gid
+            end
           end
 
-          # First mount command uses getent to get the group
-          mount_options = "-o uid=#{mount_uid},gid=#{mount_gid}"
-          mount_options += ",#{options[:mount_options].join(",")}" if options[:mount_options]
-          mount_commands << "mount -t vboxsf #{mount_options} #{name} #{guest_path}"
-
-          # Second mount command uses the old style `id -g`
-          mount_options = "-o uid=#{mount_uid},gid=#{mount_gid_old}"
-          mount_options += ",#{options[:mount_options].join(",")}" if options[:mount_options]
-          mount_commands << "mount -t vboxsf #{mount_options} #{name} #{guest_path}"
+          mount_options = options.fetch(:mount_options, [])
```

Comment:
The change of the folder option ordering introduces the issue related to #7859.

E.g:
Old ordering
`mount -t vboxsf -o uid=1001,gid=1001,uid=201,gid=201 foo /foo`

New ordering
`mount -t vboxsf -o uid=201,gid=201,uid=1001,gid=1001 foo /foo`